File: packages/python-sdk/src/authsec_sdk/runtime/validator.py
"""HybridValidator — JWT verification + RFC 7662 introspection.

Mirrors :type:`authsec.HybridValidator` in the Go SDK. Supports four
:class:`ValidationMode` shapes; the strict ``JWT_AND_INTROSPECT`` mode is the
production-recommended setting.

JWT verification uses cached JWKS keys, refreshed on `kid` miss. Introspection
uses HTTP Basic auth with the configured ``introspection_client_id`` /
``introspection_client_secret``.
"""

from __future__ import annotations

import time
from typing import Any, Optional

import aiohttp
import jwt
from jwt import PyJWKClient

from .config import Config, ValidationMode
from .principal import Principal


class TokenInvalidError(Exception):
    """Raised when the token fails validation (signature, expiry, audience, …)."""


class TokenInactiveError(Exception):
    """Raised when introspection returns ``active=false`` (revoked / suspended)."""


class HybridValidator:
    """Validates bearer tokens via JWT + introspection.

    Construct with :func:`new_validator`; the constructor configures HTTP
    clients, the JWKS cache, and validates the config against the chosen mode.

    Validation flow per mode:

    JWT_ONLY
        Verify signature + claims locally via JWKS. No backend call.

    INTROSPECTION_ONLY
        POST to ``introspection_url``. Token is whatever the AS returns.

    JWT_AND_INTROSPECT (recommended)
        For JWT-shaped tokens, JWT verification must succeed AND introspection
        must return ``active=true``. For opaque tokens, introspection alone.
        Both failures are terminal.

    JWT_OR_INTROSPECT
        Either path may succeed independently. Use only during gradual migration.
    """

    def __init__(self, cfg: Config) -> None:
        self.cfg = cfg.normalized()
        self._mode = self.cfg.effective_validation_mode()
        self._jwks_client: Optional[PyJWKClient] = None
        if self.cfg.jwks_url.strip():
            # PyJWKClient caches per-kid keys and refreshes on miss.
            self._jwks_client = PyJWKClient(
                self.cfg.jwks_url,
                cache_keys=True,
                lifespan=3600,
            )

    # ────────────────────────────────────────────────────────────
    # Public entry point
    # ────────────────────────────────────────────────────────────

    async def validate(self, token: str) -> Principal:
        """Validate ``token`` and return a populated :class:`Principal`.

        Raises :class:`TokenInvalidError` if validation fails or
        :class:`TokenInactiveError` if introspection signals inactive.
        """
        mode = self._mode
        if mode == ValidationMode.JWT_ONLY:
            return self._validate_jwt(token)
        if mode == ValidationMode.INTROSPECTION_ONLY:
            return self._check_active(await self._introspect(token))
        if mode == ValidationMode.JWT_AND_INTROSPECT:
            return await self._validate_jwt_and_introspect(token)
        if mode == ValidationMode.JWT_OR_INTROSPECT:
            return await self._validate_jwt_or_introspect(token)
        raise TokenInvalidError(f"unsupported validation mode: {mode}")

    # ────────────────────────────────────────────────────────────
    # Mode implementations
    # ────────────────────────────────────────────────────────────

    async def _validate_jwt_and_introspect(self, token: str) -> Principal:
        """Strict: JWT must pass, then introspection must return active.

        For opaque (non-JWT) tokens, JWT verification is skipped and
        introspection alone decides.
        """
        looks_like_jwt = token.count(".") == 2
        jwt_principal: Optional[Principal] = None
        if looks_like_jwt:
            jwt_principal = self._validate_jwt(token)

        introspected = await self._introspect(token)
        active = self._check_active(introspected)

        if jwt_principal is None:
            return active
        return self._merge_principals(jwt_principal, active)

    async def _validate_jwt_or_introspect(self, token: str) -> Principal:
        """Either succeeds independently (legacy compatibility)."""
        jwt_err: Optional[Exception] = None
        try:
            return self._validate_jwt(token)
        except Exception as e:  # noqa: BLE001 — we deliberately catch every JWT failure
            jwt_err = e

        try:
            return self._check_active(await self._introspect(token))
        except Exception as introspect_err:
            raise TokenInvalidError(
                f"JWT verify failed ({jwt_err}); introspection failed ({introspect_err})"
            ) from introspect_err

    # ────────────────────────────────────────────────────────────
    # JWT path
    # ────────────────────────────────────────────────────────────

    def _validate_jwt(self, token: str) -> Principal:
        if self._jwks_client is None:
            raise TokenInvalidError("JWT verification requested but no jwks_url configured")
        try:
            signing_key = self._jwks_client.get_signing_key_from_jwt(token).key
        except Exception as e:  # JWK lookup / key parse failure
            raise TokenInvalidError(f"could not resolve signing key: {e}") from e

        try:
            decoded = jwt.decode(
                token,
                signing_key,
                algorithms=["RS256", "RS384", "RS512", "ES256", "ES384"],
                # We deliberately verify aud in the higher-level
                # request-handling layer (against resource_uri) rather than
                # via PyJWT's audience matcher, so we can produce richer
                # error messages.
                options={"verify_aud": False, "verify_iss": True},
                issuer=self.cfg.issuer or None,
            )
        except jwt.ExpiredSignatureError as e:
            raise TokenInvalidError("token expired") from e
        except jwt.InvalidIssuerError as e:
            raise TokenInvalidError(f"invalid issuer: {e}") from e
        except jwt.InvalidTokenError as e:
            raise TokenInvalidError(f"invalid JWT: {e}") from e

        return Principal(
            subject=_str(decoded.get("sub")),
            issuer=_str(decoded.get("iss")),
            audience=_audience_list(decoded.get("aud")),
            scopes=_scope_list(decoded.get("scope")) or _scope_list(decoded.get("scopes")),
            claims=decoded,
            active=True,
        )

    # ────────────────────────────────────────────────────────────
    # Introspection path
    # ────────────────────────────────────────────────────────────

    async def _introspect(self, token: str) -> Principal:
        if not self.cfg.introspection_url.strip():
            raise TokenInvalidError("introspection requested but no introspection_url configured")

        auth = aiohttp.BasicAuth(
            self.cfg.introspection_client_id, self.cfg.introspection_client_secret
        )
        timeout = aiohttp.ClientTimeout(total=self.cfg.request_timeout_seconds)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(
                self.cfg.introspection_url,
                data={"token": token},
                auth=auth,
                headers={"Accept": "application/json"},
            ) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise TokenInvalidError(
                        f"introspection returned HTTP {resp.status}: {body[:200]}"
                    )
                payload = await resp.json()

        return Principal(
            subject=_str(payload.get("sub") or payload.get("subject")),
            issuer=_str(payload.get("iss")),
            audience=_audience_list(payload.get("aud") or payload.get("resource")),
            scopes=_scope_list(payload.get("scope")) or _scope_list(payload.get("scopes")),
            claims=dict(payload),
            active=bool(payload.get("active", False)),
        )

    # ────────────────────────────────────────────────────────────
    # Helpers
    # ────────────────────────────────────────────────────────────

    def _check_active(self, principal: Principal) -> Principal:
        if not principal.active:
            raise TokenInactiveError(
                "introspection returned active=false — token revoked, "
                "suspended, expired, or principal membership inactive"
            )
        return principal

    def _merge_principals(self, jwt_p: Principal, introspected: Principal) -> Principal:
        """Combine a JWT-derived principal with the introspection result.

        Introspection is authoritative for ``active`` and scope set; JWT
        contributes claims that may not be re-emitted by introspection.
        """
        merged_claims = dict(jwt_p.claims)
        merged_claims.update(introspected.claims)
        return Principal(
            subject=introspected.subject or jwt_p.subject,
            issuer=introspected.issuer or jwt_p.issuer,
            audience=introspected.audience or jwt_p.audience,
            scopes=introspected.scopes or jwt_p.scopes,
            claims=merged_claims,
            active=introspected.active,
        )


# ────────────────────────────────────────────────────────────────────
# Module-level helpers (mirror Go's stringClaim / audienceFromClaims / etc.)
# ────────────────────────────────────────────────────────────────────


def _str(value: Any) -> str:
    if isinstance(value, str):
        return value
    return ""


def _audience_list(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, str):
        return [value]
    if isinstance(value, (list, tuple)):
        return [str(v) for v in value if v]
    return [str(value)]


def _scope_list(value: Any) -> list[str]:
    """Accept space-separated string OR list[str]."""
    if value is None:
        return []
    if isinstance(value, str):
        return [s for s in value.split() if s]
    if isinstance(value, (list, tuple)):
        return [str(s) for s in value if s]
    return []


def new_validator(cfg: Config) -> HybridValidator:
    """Construct a :class:`HybridValidator` after validating the config."""
    cfg.validate()
    return HybridValidator(cfg)

File: packages/typescript-sdk/src/runtime/validator.ts
/**
 * HybridValidator — JWT verification + RFC 7662 introspection.
 *
 * TypeScript parity port of python/runtime/validator.py.
 *
 * Supports four ValidationMode shapes; ``jwt_and_introspect`` is the
 * production-recommended setting. JWT verification uses cached JWKS keys via
 * the ``jose`` package; introspection uses HTTP Basic auth with the
 * configured introspectionClientId / introspectionClientSecret.
 */

import type { JWTPayload, JWTVerifyGetKey } from 'jose' with { 'resolution-mode': 'import' };

import {
  effectiveValidationMode,
  normalizeConfig,
  validateConfig,
  type Config,
  type ValidationMode,
} from './config.js';
import { newPrincipal, type Principal } from './principal.js';

// `jose` is ESM-only. We import it lazily so this module can be consumed
// from both CommonJS and ESM TS configurations without a top-level await.
type JoseModule = typeof import('jose', { with: { 'resolution-mode': 'import' } });
let _josePromise: Promise<JoseModule> | null = null;
async function loadJose(): Promise<JoseModule> {
  if (_josePromise === null) {
    _josePromise = import('jose') as Promise<JoseModule>;
  }
  return _josePromise;
}

/** Denial-taxonomy code attached to ValidationError. */
export type ValidationErrorCode =
  | 'invalid_token'
  | 'invalid_audience'
  | 'token_inactive';

/** Error raised when the token fails validation or introspection. */
export class ValidationError extends Error {
  readonly code: ValidationErrorCode;
  constructor(code: ValidationErrorCode, message: string) {
    super(message);
    this.name = 'ValidationError';
    this.code = code;
  }
}

/**
 * Validates bearer tokens via JWT + introspection.
 *
 * Construct with ``newValidator()`` which validates the config first, or
 * directly via ``new HybridValidator(cfg)``.
 */
export class HybridValidator {
  readonly cfg: Config;
  private readonly mode: ValidationMode;
  private jwks: JWTVerifyGetKey | null = null;
  private jwksInit: Promise<void> | null = null;

  constructor(cfg: Config) {
    this.cfg = normalizeConfig(cfg);
    this.mode = effectiveValidationMode(this.cfg);
  }

  private async ensureJwks(): Promise<JWTVerifyGetKey | null> {
    if (!this.cfg.jwksUrl.trim()) return null;
    if (this.jwks !== null) return this.jwks;
    if (this.jwksInit === null) {
      this.jwksInit = (async () => {
        const jose = await loadJose();
        this.jwks = jose.createRemoteJWKSet(new URL(this.cfg.jwksUrl), {
          cacheMaxAge: 3600 * 1000,
          cooldownDuration: 30 * 1000,
        });
      })();
    }
    await this.jwksInit;
    return this.jwks;
  }

  /**
   * Validate ``token`` and return a populated Principal.
   *
   * Throws ``ValidationError`` with ``.code`` ∈ ``invalid_token`` /
   * ``token_inactive`` on failure.
   */
  async validate(token: string): Promise<Principal> {
    switch (this.mode) {
      case 'jwt_only':
        return await this.validateJwt(token);
      case 'introspection_only': {
        const p = await this.introspect(token);
        return this.checkActive(p);
      }
      case 'jwt_and_introspect':
        return await this.validateJwtAndIntrospect(token);
      case 'jwt_or_introspect':
        return await this.validateJwtOrIntrospect(token);
      default:
        throw new ValidationError('invalid_token', `unsupported validation mode: ${this.mode}`);
    }
  }

  // ── Mode implementations ──────────────────────────────────────────

  private async validateJwtAndIntrospect(token: string): Promise<Principal> {
    const looksLikeJwt = token.split('.').length === 3;
    let jwtPrincipal: Principal | null = null;
    if (looksLikeJwt) {
      jwtPrincipal = await this.validateJwt(token);
    }
    const introspected = await this.introspect(token);
    const active = this.checkActive(introspected);
    if (jwtPrincipal === null) return active;
    return this.mergePrincipals(jwtPrincipal, active);
  }

  private async validateJwtOrIntrospect(token: string): Promise<Principal> {
    let jwtErr: unknown;
    try {
      return await this.validateJwt(token);
    } catch (e) {
      jwtErr = e;
    }
    try {
      const p = await this.introspect(token);
      return this.checkActive(p);
    } catch (introspectErr) {
      const jm = jwtErr instanceof Error ? jwtErr.message : String(jwtErr);
      const im = introspectErr instanceof Error ? introspectErr.message : String(introspectErr);
      throw new ValidationError(
        'invalid_token',
        `JWT verify failed (${jm}); introspection failed (${im})`,
      );
    }
  }

  // ── JWT path ───────────────────────────────────────────────────────

  private async validateJwt(token: string): Promise<Principal> {
    const jwks = await this.ensureJwks();
    if (jwks === null) {
      throw new ValidationError(
        'invalid_token',
        'JWT verification requested but no jwksUrl configured',
      );
    }
    const jose = await loadJose();
    let payload: JWTPayload;
    try {
      const result = await jose.jwtVerify(token, jwks, {
        // We deliberately verify aud in the higher-level request-handling
        // layer (Runtime.validateToken) against resourceUri, so we can
        // produce richer error messages.
        issuer: this.cfg.issuer || undefined,
        algorithms: ['RS256', 'RS384', 'RS512', 'ES256', 'ES384'],
      });
      payload = result.payload;
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      // jose throws JWTExpired with code 'ERR_JWT_EXPIRED'
      if (e && typeof e === 'object' && 'code' in e && (e as { code?: string }).code === 'ERR_JWT_EXPIRED') {
        throw new ValidationError('invalid_token', 'token expired');
      }
      throw new ValidationError('invalid_token', `invalid JWT: ${msg}`);
    }
    return newPrincipal({
      subject: asString(payload.sub),
      issuer: asString(payload.iss),
      audience: audienceList(payload.aud),
      scopes:
        scopeList((payload as Record<string, unknown>).scope) ||
        scopeList((payload as Record<string, unknown>).scopes),
      claims: payload as Record<string, unknown>,
      active: true,
    });
  }

  // ── Introspection path ────────────────────────────────────────────

  private async introspect(token: string): Promise<Principal> {
    if (!this.cfg.introspectionUrl.trim()) {
      throw new ValidationError(
        'invalid_token',
        'introspection requested but no introspectionUrl configured',
      );
    }
    const credentials = Buffer.from(
      `${this.cfg.introspectionClientId}:${this.cfg.introspectionClientSecret}`,
    ).toString('base64');

    const controller = new AbortController();
    const timer = setTimeout(
      () => controller.abort(),
      Math.max(1, this.cfg.requestTimeoutSeconds) * 1000,
    );
    let resp: Response;
    try {
      resp = await fetch(this.cfg.introspectionUrl, {
        method: 'POST',
        headers: {
          Authorization: `Basic ${credentials}`,
          Accept: 'application/json',
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: new URLSearchParams({ token }).toString(),
        signal: controller.signal,
      });
    } catch (e) {
      throw new ValidationError(
        'invalid_token',
        `introspection request failed: ${e instanceof Error ? e.message : String(e)}`,
      );
    } finally {
      clearTimeout(timer);
    }
    if (resp.status !== 200) {
      const body = await resp.text().catch(() => '');
      throw new ValidationError(
        'invalid_token',
        `introspection returned HTTP ${resp.status}: ${body.slice(0, 200)}`,
      );
    }
    let payload: Record<string, unknown>;
    try {
      payload = (await resp.json()) as Record<string, unknown>;
    } catch (e) {
      throw new ValidationError(
        'invalid_token',
        `introspection returned non-JSON body: ${e instanceof Error ? e.message : String(e)}`,
      );
    }
    return newPrincipal({
      subject: asString(payload.sub ?? payload.subject),
      issuer: asString(payload.iss),
      audience: audienceList(payload.aud ?? payload.resource),
      scopes: scopeList(payload.scope) || scopeList(payload.scopes),
      claims: { ...payload },
      active: Boolean(payload.active),
    });
  }

  // ── Helpers ────────────────────────────────────────────────────────

  private checkActive(principal: Principal): Principal {
    if (!principal.active) {
      throw new ValidationError(
        'token_inactive',
        'introspection returned active=false — token revoked, suspended, expired, or principal membership inactive',
      );
    }
    return principal;
  }

  private mergePrincipals(jwtP: Principal, introspected: Principal): Principal {
    return {
      subject: introspected.subject || jwtP.subject,
      issuer: introspected.issuer || jwtP.issuer,
      audience: introspected.audience.length > 0 ? introspected.audience : jwtP.audience,
      scopes: introspected.scopes.length > 0 ? introspected.scopes : jwtP.scopes,
      claims: { ...jwtP.claims, ...introspected.claims },
      active: introspected.active,
    };
  }
}

// ── Module-level helpers ─────────────────────────────────────────────

function asString(v: unknown): string {
  return typeof v === 'string' ? v : '';
}

function audienceList(v: unknown): string[] {
  if (v === null || v === undefined) return [];
  if (typeof v === 'string') return [v];
  if (Array.isArray(v)) return v.filter(Boolean).map(String);
  return [String(v)];
}

function scopeList(v: unknown): string[] {
  if (v === null || v === undefined) return [];
  if (typeof v === 'string') return v.split(/\s+/).filter(Boolean);
  if (Array.isArray(v)) return v.filter(Boolean).map(String);
  return [];
}

/** Construct a HybridValidator after validating the config. */
export function newValidator(cfg: Config): HybridValidator {
  validateConfig(cfg);
  return new HybridValidator(cfg);
}
